canonical/main.go

```diff
@@ -0,0 +1,377 @@
+
+func main() {
+	inBucket := flag.String("inbucket", "", "the input bucket")
+	outBucket := flag.String("outbucket", "", "the output bucket")
+	processEverything := flag.Bool("f", false, "process every tarball")
+
+	cpuprofile := flag.String("cpuprofile", "", "write cpu profile to file")
+	memprofile := flag.String("memprofile", "", "write memory profile to this file")
+
+	flag.Parse()
+
+	if *inBucket == "" {
+		log.Fatal("Must provide -inbucket")
+	}
+	if *outBucket == "" {
+		log.Fatal("Must provide -outbucket")
+	}
+	if *inBucket == *outBucket {
+		log.Fatal("-inbucket must be different from -outbucket")
+	}
+
+	if *cpuprofile != "" {
+		f, err := os.Create(*cpuprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	err := processBucket(ctx, *inBucket, *outBucket, *processEverything)
+
+	if *memprofile != "" {
+		f, err := os.Create(*memprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		runtime.GC() // get up-to-date statistics
+		pprof.WriteHeapProfile(f)
+		f.Close()
+		return
+	}
+
+	if err != nil {
+		fmt.Println("error processing bucket:", err)
+		return
+	}
+}
+
+// processBucket processes all tarballs in the given bucket
+// and stores the output in the output bucket.
+// If processEverything is false, only new tarballs will be processed.
+func processBucket(ctx context.Context, inBucketName, outBucketName string, processEverything bool) error {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+
+	inBucket := client.Bucket(inBucketName)
+	if _, err := inBucket.Attrs(ctx); err != nil {
+		return fmt.Errorf("could not get bucket %q: %v", inBucketName, err)
+	}
+
+	outBucket := client.Bucket(outBucketName)
+	if _, err := outBucket.Attrs(ctx); err != nil {
+		return fmt.Errorf("could not get bucket %q: %v", outBucketName, err)
+	}
+
+	allInputs, err := objects(ctx, inBucket)
+	if err != nil {
+		return err
+	}
+
+	existingOutputs, err := objects(ctx, outBucket)
+	if err != nil {
+		return err
+	}
+
+	inputs := []string{}
+	for name := range allInputs {
+		if processEverything || !existingOutputs[name] {
+			inputs = append(inputs, name)
+		}
+	}
+
+	if len(inputs) == 0 {
+		fmt.Println("Nothing to do. Add -f flag?")
+		return nil
+	}
+
+	bar := pbTemplate.Start(len(inputs)).Set("prefix", "Canonicalizing... ").SetRefreshRate(time.Second)
+
+	g, ctx := errgroup.WithContext(ctx)
+	sem := semaphore.NewWeighted(500)
+
+	for _, name := range inputs {
+		if err := ctx.Err(); ctx.Err() != nil {
+			bar.Finish()
+			log.Println(err)
+			break
+		}
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			bar.Finish()
+			log.Println(err)
+			break
+		}
+		releaseOnce := &sync.Once{}
```

Comment:
Why is this needed?

Reply:
Removed. Was seeing a panic in some ctrl+c situations, but don't see it now.